pkg/router/controller/router_controller.go

```diff
@@ -132,7 +132,11 @@ func (c *RouterController) processNamespace(eventType watch.EventType, ns *kapi.
 }
 
 func (c *RouterController) UpdateNamespaces() {
-	namespaces := c.FilteredNamespaceNames
+	// Note: Need to clone the filtered namespace names or else any updates
+	//       we make locally in processNamespace() will be immediately
+	//       reflected to plugins in the chain beneath us. This creates
+	//       cleanup issues as old == new in Plugin.HandleNamespaces().
+	namespaces := sets.NewString(c.FilteredNamespaceNames.List()...)
```

Comment:
Oof, good catch

Reply:
Yes, processNamespace() is synchronous but Host admitter and Unique status plugins are holding reference to the FilteredNamespaceNames and any subsequent changes in processNamespace() are immediately reflected in those plugin.